packages/aws-cdk-lib/aws-cloudfront-origins/lib/s3-bucket-origin.ts

```diff
@@ -61,10 +70,27 @@ export abstract class S3BucketOrigin extends cloudfront.OriginBase {
   public static withOriginAccessControl(bucket: IBucket, props?: S3BucketOriginWithOACProps): cloudfront.IOrigin {
     return new class extends S3BucketOrigin {
       private originAccessControl?: cloudfront.IOriginAccessControl;
+      private assembleDomainName?: boolean
 
       constructor() {
         super(bucket, { ...props });
         this.originAccessControl = props?.originAccessControl;
+        this.assembleDomainName = props?.assembleDomainName ?? false;
+
+        if (this.assembleDomainName) {
+          let bucketName: string | undefined = undefined;
+          if (!Token.isUnresolved(bucket.bucketName)) {
+            // this is the case when bucket is from Bucket.fromBucketName
+            bucketName = bucket.bucketName;
+          } else if (!Token.isUnresolved((bucket.node.defaultChild as CfnBucket)?.bucketName)) {
+            // this is the case when bucket is a L2 bucket with bucketName set
+            bucketName = (bucket.node.defaultChild as CfnBucket)?.bucketName;
+          }
+
+          if (!bucketName) {
+            throw new Error(`Cannot assemble static DomainName as bucket ${bucket.node.id} has no bucketName set.`);
```

Comment:
I see it's getting the bucket name using 
```
${(bucket.node.defaultChild as CfnBucket).bucketName}
```
at Line 132. Do we still need to check `Token.isUnresolved(bucket.bucketName)`?

Reply:
Good catch! We do. Thank you.

I will cache the `bucketName` in the class and reuse it at Line 132.

I will change this PR to `Draft` as well since this approach is pending security discussion.